File: jupyter2/problems/minc/minc/go/orig/ast.go
package main

// type_expr = "long"
type TypeExpr int

const (
	TyLong TypeExpr = iota
)

// unary op: + - ! ~
type UnaryOp int

const (
	OpPos  UnaryOp = iota // +
	OpNeg                 // -
	OpLNot                // !
	OpBNot                // ~
)

// binary op: + - * / % < > <= >= == !=
type BinaryOp int

const (
	OpAdd BinaryOp = iota // +
	OpSub                 // -
	OpMul                 // *
	OpDiv                 // /
	OpMod                 // %
	OpLt                  // <
	OpGt                  // >
	OpLe                  // <=
	OpGe                  // >=
	OpEq                  // ==
	OpNe                  // !=
)

// expr
type Expr interface{ isExpr() }

type IntLit struct{ Value int64 } // number  123
type VarRef struct{ Name string } // identifier  x
type Call struct {                // call  f(x, y)
	Name string
	Args []Expr
}
type Unary struct { // unary  -x
	Op UnaryOp
	E  Expr
}
type Binary struct { // binary  x + y
	Op   BinaryOp
	L, R Expr
}
type Assign struct { // assignment  x = e
	LHS Expr
	RHS Expr
}

func (*IntLit) isExpr() {}
func (*VarRef) isExpr() {}
func (*Call) isExpr()   {}
func (*Unary) isExpr()  {}
func (*Binary) isExpr() {}
func (*Assign) isExpr() {}

// var_decl  "long x;"
type VarDecl struct {
	Type TypeExpr
	Name string
}

// stmt
type Stmt interface{ isStmt() }

type EmptyStmt struct{}          // ;
type ContinueStmt struct{}       // continue;
type BreakStmt struct{}          // break;
type ReturnStmt struct{ E Expr } // return e;
type Compound struct {           // { var_decl* stmt* }
	Decls []VarDecl
	Stmts []Stmt
}
type IfStmt struct { // if (e) s [else s]
	Cond Expr
	Then Stmt
	Else Stmt // nil if there is no else
}
type ExprStmt struct{ E Expr } // e;

func (*EmptyStmt) isStmt()    {}
func (*ContinueStmt) isStmt() {}
func (*BreakStmt) isStmt()    {}
func (*ReturnStmt) isStmt()   {}
func (*Compound) isStmt()     {}
func (*IfStmt) isStmt()       {}
func (*ExprStmt) isStmt()     {}

// parameter  "long x"
type Param struct {
	Type TypeExpr
	Name string
}

// fun_definition  "long f(long x, long y) { ... }"
type FunDef struct {
	RetType TypeExpr
	Name    string
	Params  []Param
	Body    *Compound
}

// program = {definition}*
type Program struct {
	Funcs []*FunDef
}

File: jupyter2/problems/minc/minc/ml/orig/lib/ast.ml
(* type_expr = "long" *)
type type_expr =
  | TyLong

(* unary_op: ("+"|"-"|"!"|"~") unary_expr *)
type unary_op =
  | Pos    (* + *)
  | Neg    (* - *)
  | LNot   (* ! *)
  | BNot   (* ~ *)

(* binary_op *)
type binary_op =
  | Add | Sub             (* + - *)
  | Mul | Div | Mod       (* * / % *)
  | Lt | Gt | Le | Ge     (* < > <= >= *)
  | Eq | Ne               (* == != *)

(* expr *)
type expr =
  | Int    of int                  (* 123 *)
  | Var    of string               (* x *)
  | Call   of string * expr list   (* f(x, y) *)
  | Unary  of unary_op * expr      (* -x *)
  | Binary of binary_op * expr * expr  (* x + y *)
  | Assign of expr * expr          (* x = e *)

(* var_decl = "long" identifier ";" *)
type var_decl = {
  vd_type : type_expr;
  vd_name : string;
}

(* stmt *)
type stmt =
  | Empty                              (* ; *)
  | Continue                           (* continue; *)
  | Break                              (* break; *)
  | Return   of expr                   (* return e; *)
  | Compound of var_decl list * stmt list  (* { var_decl* stmt* } *)
  | If       of expr * stmt * stmt option  (* if (e) s [else s] *)
  | Expr     of expr                   (* e; *)

(* parameter = "long" identifier *)
type parameter = {
  p_type : type_expr;
  p_name : string;
}

(* fun_definition = "long" f "(" parameter_list ")" compound_stmt *)
type fun_definition = {
  f_ret    : type_expr;
  f_name   : string;
  f_params : parameter list;
  f_body   : stmt;
}

(* definition = fun_definition *)
type definition =
  | FunDef of fun_definition

(* program = {definition}* *)
type program = definition list
